tests/utils.py

```diff
@@ -31,3 +32,50 @@ async def wrapper(*args, **kwargs):
         return result
 
     return wrapper
+
+
+@contextlib.contextmanager
+def mock_transport_responses(transport, responses):
+    """
+    Mocks the responses in the provided HTTPX transport by
+    shadowing the request handlers just before the HTTPTransport
+    or AsyncHTTPTransport in the transport's MRO.
+
+    Returns a class from which some metadata about the mocking can
+    be read (such as number of requests).
+    """
+    responses = iter(responses)
+
+    class ResponseMetaData:
+        num_requests = 0
+
+    class MockTransport:
+        @staticmethod
+        def handle_request(request, *_, **__):
+            try:
+                ResponseMetaData.num_requests += 1
+                response = responses.__next__()
+                response.request = request
+                return response
+            except StopIteration as e:
+                raise RuntimeError("Ran out of mock responses") from e
+
+        async def handle_async_request(self, *args, **kwargs):
+            return self.handle_request(*args, **kwargs)
+
```

Comment:
Did you find this doc-entry: https://www.python-httpx.org/advanced/#mock-transports ?

It points to a built-in solution which probably is more robust down the line, but also points to fully-fledged mocking solutions for, for example, pytest, which I guess would be even handier. Typically, you would need to rewrite your test to query different virtual endpoints, but I think that would be worthwhile for a more robust solution?!

Reply:
No probs... if there is a reason to do this, it is all fine!